pkg/workloads/spark_job/spark_util.py

```diff
@@ -246,26 +258,40 @@ def ingest(ctx, spark):
 
 def read_csv(ctx, spark):
     data_config = ctx.environment["data"]
-    expected_field_names = data_config["schema"]
-
-    schema_fields = []
-    for field_name in expected_field_names:
-        if field_name in ctx.raw_columns:
-            spark_type = CORTEX_TYPE_TO_SPARK_TYPE[ctx.raw_columns[field_name]["type"]]
-        else:
-            spark_type = StringType()
-
-        schema_fields.append(StructField(name=field_name, dataType=spark_type))
 
     csv_config = {
         util.snake_to_camel(param_name): val
         for param_name, val in data_config.get("csv_config", {}).items()
         if val is not None
     }
 
-    df = spark.read.csv(
-        data_config["path"], schema=StructType(schema_fields), mode="FAILFAST", **csv_config
-    )
+    df = spark.read.csv(data_config["path"], inferSchema=True, **csv_config)
+    renamed_cols = [F.col(c).alias(data_config["schema"][idx]) for idx, c in enumerate(df.columns)]
+    df = df.select(*renamed_cols)
+
+    if set(data_config["schema"]) != set(df.columns):
+        logger.error("found schema:")
+        log_df_schema(df, logger.error)
+        raise UserException(
+            "expected column(s) "
+            + str(set(data_config["schema"]))
+            + " but got "
+            + str(set(df.columns))
+        )
+
+    # for columns with types, cast it
+    casted_cols = []
+    for column_name in ctx.raw_columns:
+        column_type = ctx.raw_columns[column_name]["type"]
+        if column_type == consts.COLUMN_TYPE_INFERRED:
+            casted_cols.append(F.col(column_name).alias(column_name))
```

Comment:
alias is not needed here since no transformation has been applied to the column
`casted_cols.append(F.col(column_name))`